pkg/dyninst/irgen/irgen.go

```diff
@@ -196,101 +210,68 @@ func generateIR(
 	}, nil
 }
 
-type processDwarfResult struct {
-	pendingSubprograms []*pendingSubprogram
-	firstAbstractIdx   int
+type prologueSeachParams struct {
+	unit    *dwarf.Entry
+	pcRange ir.PCRange
 }
 
-// prologueResult captures the outcome of prologue discovery for a given
-// subprogram. If err is non-nil, the location should be considered invalid.
-type prologueResult struct {
-	loc ir.InjectionPoint
-	err error
+type prologueEndLocation struct {
+	err       error
+	frameless bool
+	// If not frameless, the pc of the prologue end, otherwise the first pc
+	// of the subprogram.
+	pc uint64
 }
 
-// computePrologueResults walks each interesting sub-program and determines the
-// appropriate injection point at the end of its prologue. The returned slice is
-// indexed the same way as the pending slice.
-func computePrologueResults(
+// findPrologues searches for prologue for each given search param. Results
+// are indexed by the start of the pc range. Provided ranges must be non-overlapping
+// but may contain duplicates.
+func findProloguesEnds(
 	objFile object.File,
-	pending []*pendingSubprogram,
-	abstractStartIdx int,
-) ([]prologueResult, error) {
-	arch := objFile.Architecture()
-
-	ranks := make([]uint32, len(pending))
-	for i := range pending {
-		ranks[i] = uint32(i)
-	}
-
-	getPC := func(i uint32) uint64 {
-		if ool := pending[i].subprogram.OutOfLinePCRanges; len(ool) > 0 {
-			return ool[0][0]
-		}
-		return math.MaxUint64
-	}
-
-	slices.SortFunc(ranks, func(a, b uint32) int {
-		return cmp.Or(
-			cmp.Compare(pending[a].unit.Offset, pending[b].unit.Offset),
-			cmp.Compare(getPC(a), getPC(b)),
-		)
+	searchParams []prologueSeachParams,
+) (map[uint64]prologueEndLocation, error) {
+	slices.SortFunc(searchParams, func(a, b prologueSeachParams) int {
+		return cmp.Compare(a.pcRange[0], b.pcRange[0])
 	})
 
-	results := make([]prologueResult, len(pending))
+	res := make(map[uint64]prologueEndLocation)
```

Comment:
nit: if you compact the ranges to eliminate duplicates, you can pass the correct size to this make call to avoid reallocations.